### DISCUSSION TITLE
Split `DataTypes.sol` in separate files

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#488] [Ideas]
This idea just came to me - what if we split `DataTypes.sol` into multiple files?

- `types/Generics.sol`
- `types/Lockup.sol`
- `types/LockupDynamic.sol`
- `types/LockupLinear.sol`

This design is (i) more specific, (ii) less repetitive, and (ii) more robust in the long term because the `DataTypes.sol` file would not get cluttered with increasingly more types.

**Comment by andreivladbrg** [comment]
> This design is (i) more specific, (ii) less repetitive, and (ii) more robust in the long term 

I am sorry I haven't seen your discussion earlier, but I don't agree with your points.

i) A well-named and well-organized single file can be just as specific. Moreover, having them in one file allows developers to get an overview of all the data types at once(better DX), which is often more beneficial for understanding the system as a whole.

ii) Why this is not true: The constant multiple imports that someone has to do would simply not be "less repetitive". This method would likely increase the amount of repetitive code, as each file will have to import its dependencies independently. Moreover, it make the structure more complex to manage.

iii) Splitting the data types into separate files could increase the surface area for mistakes and misunderstandings, particularly if the separation makes it harder to see how the types are connected to each other.

The advantages of having all your structs and libraries under the same umbrella:

- Easier Navigation: Developers can see all the related libraries and structs in one place, making it easier to understand the relationships and dependencies between them.
- Decreased Complexity: The cognitive load is decreased when you have a single file to deal with, as opposed to navigating through a multitude of files. This, in turn, can improve the productivity.
- Reduced Risk of Errors: With fewer files and less repetition of import statements, there's a lower risk of making errors.

Given these considerations, I strongly suggest for maintaining our current approach.

**Reply by PaulRBerg** [reply]
I agree with the general position of your reply, as well as your conclusion. Whatever minor disagreements I had, I replied below:

> A well-named and well-organized single file can be just as specific

You used this term differently - by "specific", I meant that each library belongs to a dedicated file.

> The constant multiple imports that someone has to do would simply not be "less repetitive".

I meant that the word "types" would not be duplicated in the path names, i.e., `types/DataTypes` ("types" is used twice in this path).

---

Final comment: one other benefit of splitting the types as I did is that the compilation times are faster if we separate the types. But this is a negligible benefit, and we can never mind it.
